### PR TITLE
WINC-476 : [wmco] Bump operator-sdk to v0.19.4

### PR DESCRIPTION
**Pull request by VaishnaviHire**
This commit upgrades operator-sdk to v0.19.4. Readme and all
other SDK dependencies are bumped to a supported version.

[WINC-476](https://issues.redhat.com/browse/WINC-476)

Commands that are deprecated but supported in v0.19.4 -

- `generate csv`
- `add controller`
- `new project`

### CODE REVIEW

**Comment by mansikulkarni96** [comment]
LGTM

**Comment by sebsoto** [comment]
@VaishnaviHire There are CI failures due to operator deployment, are they related to the upgrade?

File: vendor/github.com/golang/protobuf/proto/buffer.go [hunk 1]
```diff
@@ -0,0 +1,324 @@
+// Copyright 2019 The Go Authors. All rights reserved.
```
Comment (sebsoto) [inline vendor/github.com/golang/protobuf/proto/buffer.go:1]: vendoring should be seperated into its own commit

**Review by sebsoto — CHANGES_REQUESTED** [review]

**Comment by LorbusChris** [comment]
` time="2020-10-06T21:30:38Z" level=fatal msg="Failed to run operator: error waiting for CSV to install: csv failed: reason: \"NoOperatorGroup\", message: \"csv in namespace with no operatorgroups\"" `

It seems to me as though nothing applies the OperatorGroup yaml right now. Should that be added here https://github.com/openshift/windows-machine-config-operator/blob/master/hack/common.sh#L37 ?

**Comment by VaishnaviHire** [comment]
> `time="2020-10-06T21:30:38Z" level=fatal msg="Failed to run operator: error waiting for CSV to install: csv failed: reason: \"NoOperatorGroup\", message: \"csv in namespace with no operatorgroups\""`
> 
> It seems to me as though nothing applies the OperatorGroup yaml right now. Should that be added here https://github.com/openshift/windows-machine-config-operator/blob/master/hack/common.sh#L37 ?

Not sure if that would help. Logs from the local run

```
INFO[0000] Running operator from directory deploy/olm-catalog/windows-machine-config-operator 
INFO[0001] A stale windows-machine-config-operator registry exists, deleting 
INFO[0001]   Deleting  "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-manifests-0-0-0" 
INFO[0001]   Deleting  "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-manifests-package" 
INFO[0001]   Deleting Deployment "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" 
INFO[0002]   Deleting Service "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" 
INFO[0002] Creating windows-machine-config-operator registry 
INFO[0002]   Creating ConfigMap "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-manifests-package" 
INFO[0002]   Creating ConfigMap "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-manifests-0-0-0" 
INFO[0002]   Creating Deployment "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" 
INFO[0003]   Creating Service "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" 
INFO[0003] Waiting for Deployment "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" rollout to complete 
INFO[0003]   Waiting for Deployment "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" to rollout: 0 of 1 updated replicas are available 
INFO[0008]   Deployment "openshift-operator-lifecycle-manager/windows-machine-config-operator-registry-server" successfully rolled out 
INFO[0008] Creating resources                           
INFO[0008]   Creating ClusterServiceVersion "windows-machine-config-operator/windows-machine-config-operator.v0.0.0" 
INFO[0008]   Creating CatalogSource "windows-machine-config-operator/windows-machine-config-operator-ocs" 
INFO[0008]   Creating Subscription "windows-machine-config-operator/windows-machine-config-operator-v0-0-0-sub" 
INFO[0008]   Creating OperatorGroup "windows-machine-config-operator/operator-sdk-og" 
INFO[0008] Waiting for ClusterServiceVersion "windows-machine-config-operator/windows-machine-config-operator.v0.0.0" to reach 'Succeeded' phase 
INFO[0008]   Found ClusterServiceVersion "windows-machine-config-operator/windows-machine-config-operator.v0.0.0" phase: Pending 
INFO[0014]   Found ClusterServiceVersion "windows-machine-config-operator/windows-machine-config-operator.v0.0.0" phase: Installing 
INFO[0026]   Found ClusterServiceVersion "windows-machine-config-operator/windows-machine-config-operator.v0.0.0" phase: Succeeded 
INFO[0027] Successfully installed "windows-machine-config-operator.v0.0.0" on OLM version "0.16.1" 
NAME                                      NAMESPACE                          KIND                     STATUS
windows-machine-config-operator.v0.0.0    windows-machine-config-operator    ClusterServiceVersion    Installed
deployment "windows-machine-config-operator" successfully rolled out
[vaishnavi@localhost windows-machine-config-operator]$ operator-sdk version
[Deprecation Notice] Operator SDK has a new CLI and project layout that is aligned with Kubebuilder.
See `operator-sdk init -h` and the following doc on how to scaffold a new project:
https://v0-19-x.sdk.operatorframework.io/docs/golang/quickstart/
To migrate existing projects to the new layout see:
https://sdk.operatorframework.io/docs/building-operators/golang/project_migration_guide/

operator-sdk version: "v0.19.4", commit: "125d0dfcc71fef4f9d7e2a42b1354cb79ffdee03", kubernetes version: "v1.18.2", go version: "go1.13.15 linux/amd64"
```

**Comment by LorbusChris** [comment]
@VaishnaviHire have you tried locally with a 4.6 cluster?

**Comment by VaishnaviHire** [comment]
> @VaishnaviHire have you tried locally with a 4.6 cluster?

Yes

**Comment by sebsoto** [comment]
@LorbusChris the operatorgroup was being created by the `operator-sdk run` command

**Comment by sebsoto** [comment]
@VaishnaviHire Looking at the logs, it seems like the operator might get installed eventually? Are we just hitting a timeout before it is?

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_windows-machine-config-operator/131/pull-ci-openshift-windows-machine-config-operator-master-aws-e2e-operator/1313582966126940160/artifacts/aws-e2e-operator/gather-extra/pods/openshift-operator-lifecycle-manager_olm-operator-c4dc8f959-fgnvt_olm-operator.log

File: go.sum [hunk 2]
```diff
@@ -1336,22 +1336,20 @@ google.golang.org/protobuf v0.0.0-20200221191635-4d8936d0db64/go.mod h1:kwYJMbMJ
...
+google.golang.org/protobuf v1.22.0/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
 google.golang.org/protobuf v1.23.0 h1:4MY060fB1DLGMB/7MBTLnwQUY6+F09GEiz6SsrNqyzM=
 google.golang.org/protobuf v1.23.0/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
-google.golang.org/protobuf v1.22.0/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
 google.golang.org/protobuf v1.23.0/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
 google.golang.org/protobuf v1.23.1-0.20200526195155-81db48ad09cc/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
 google.golang.org/protobuf v1.24.0/go.mod h1:r/3tXBNzIEhYS9I1OUVjXDlt8tc493IdKGjtUeSXeh4=
 google.golang.org/protobuf v1.25.0 h1:Ejskq+SyPohKW+1uil0JJMtmHCgJPJ/qWTxr8qp+R4c=
 google.golang.org/protobuf v1.25.0/go.mod h1:9JNX74DMeImyA3h4bdi1ymwjUzf21/xIlbajtzgsN7c=
->>>>>>> fe4c8fb...  [wmco] Bump operator-sdk to v0.19.4
```
Comment (aravindhp) [inline go.sum]: Why was this here in the first place? Looks like a bad merge.

**Review by aravindhp — APPROVED** [review]
Thanks for working on this @VaishnaviHire. Please retest `aws-e2e-operator` a few times before merging to ensure we are not introducing a flake.

/approve

**Comment by VaishnaviHire** [comment]
/test aws-e2e-operator-with-custom-vxlan-port
Failure in East-west networking

**Comment by aravindhp** [comment]
> Failure in East-west networking

In particular `TestWMCO/create/Network_validation/East_West_Networking/ip-10-0-246-46.us-west-1.compute.internal/windows_and_windows_through_a_clusterIP_svc`. If this tests is flaking we need to investigate and figure out why.

**Comment by aravindhp** [comment]
/approve cancel

We should figure out why we occasionally see `Failed to run operator: error waiting for CSV to install: csv failed: reason: \"NoOperatorGroup\", message: \"csv in namespace with no operatorgroups\"` before we can merge this.

File: hack/common.sh [hunk 3]
```diff
@@ -99,6 +90,10 @@ run_WMCO() {
 cleanup_WMCO() {
   local OSDK=$1
   # Remove the operator from openshift-windows-machine-config-operator namespace
+
+  # Revert changes made to the csv
+  sed -i "s|$OPERATOR_IMAGE|REPLACE_IMAGE|g" deploy/olm-catalog/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml
```
Comment (PratikMahajan) [inline hack/common.sh:95]: Instead of replacing every time we error out(other comments in this review), you can just catch the error of cleanup command and do the replacement after that....That would cover all the cases...

**Review by PratikMahajan — CHANGES_REQUESTED** [review]
Thanks for working on this @VaishnaviHire, PTAL at my comments...

File: hack/common.sh (hunk 3 above)
Comment (aravindhp) [reply hack/common.sh:95]: I don't follow how we would be replacing every time we error out. If we error out the first time, how will we reach the second spot where we error out?

Comment (PratikMahajan) [reply hack/common.sh:95]: In my previous comments, i had asked for reverting the `replace_image` change that we are making in run stage...  
this comment says that, instead of doing that, you can just catch `cleanup` error here and print it and do the `sed` afterwards ... 
that would cover all error out conditions

File: hack/common.sh [hunk 4]
```diff
@@ -115,3 +113,18 @@ get_version() {
 
   echo $VERSION
 }
+
+# Given the parameter, adds operator image to csv or reverts the changes made to csv
+# Parameters:
+# 1: Optional parameter with operator image value to be replaced in the csv
+replace_image() {
```
Comment (aravindhp) [inline hack/common.sh:120]: This can give rise to subtle issues if used incorrectly. Better is to make the function take in two inputs and specify the inputs i.e. `replace_image $OPERATOR_IMAGE REPLACE_IMAGE` vs `replace_image REPLACE_IMAGE $OPERATOR_IMAGE`. At which point the function can be `replace_csv_value` or something  along those lines.

File: hack/common.sh [hunk 5]
```diff
@@ -70,22 +65,18 @@ build_WMCO() {
...
-  MANIFEST_LOC=`mktemp -d`
-  trap "rm -r $MANIFEST_LOC" EXIT
-  cp -r deploy/olm-catalog/windows-machine-config-operator/ $MANIFEST_LOC
-  sed -i "s|REPLACE_IMAGE|$OPERATOR_IMAGE|g" $MANIFEST_LOC/windows-machine-config-operator/manifests/windows-machine-config-operator.clusterserviceversion.yaml
+  # add operator image to csv
+  replace_image $OPERATOR_IMAGE
 
   # Validate the operator bundle manifests
-  $OSDK bundle validate "$MANIFEST_LOC"/windows-machine-config-operator/
+  $OSDK bundle validate deploy/olm-catalog/windows-machine-config-operator/
```
Comment (aravindhp) [inline hack/common.sh:72]: Please add  `MANIFEST_LOC=deploy/olm-catalog/windows-machine-config-operator/` and use that everywhere.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: hack/common.sh [hunk 6]
```diff
@@ -115,3 +116,15 @@ get_version() {
 
   echo $VERSION
 }
+
+# Given two parameters, replaces the image value in the first parameter in the csv with the other
```
Comment (aravindhp) [inline hack/common.sh:120]: The sed command will replace any value in the CSV and not just the image value. I am fine if you just fix the comment and function name. `transform_csv`.

**Review by aravindhp — CHANGES_REQUESTED** [review]

File: hack/common.sh [hunk 7]
```diff
@@ -71,21 +69,17 @@ build_WMCO() {
 run_WMCO() {
```
Comment (PratikMahajan) [inline hack/common.sh:69]: Doc needs update

File: hack/common.sh [hunk 8]
```diff
@@ -98,8 +92,15 @@ run_WMCO() {
 # 1: path to the operator-sdk binary to use
 cleanup_WMCO() {
```
Comment (PratikMahajan) [inline hack/common.sh:93]: Doc needs update

**Review by PratikMahajan — CHANGES_REQUESTED** [review]

**Review by aravindhp — APPROVED** [review]
/approve
